File: app/src/main/java/dev/orcaxr/app/SemanticPaintPlanner.kt
package dev.orcaxr.app

import org.json.JSONArray
import org.json.JSONObject

/**
 * Smart Auto-Paint — M4: resolve a vision-LLM "paint plan" into a
 * full-coverage per-triangle slot assignment. PURE — no network, no
 * Android — so the orchestration tool stays thin and this is
 * unit-testable with a synthetic plan.
 *
 * Plan shape (what the vision model is asked to return):
 * ```
 * {
 *   "base_color": "#RRGGBB",                 // whole-model base
 *   "regions": [
 *     {"name": "stripe", "color": "#RRGGBB",
 *      "polygons": [[x1,y1,x2,y2,...], ...]}  // pixel coords in the
 *   ]                                         // rendered view's frame
 * }
 * ```
 *
 * Resolution: the base color fills every triangle (so the back / any
 * face the camera never saw is still covered — honest for a single
 * reference view); each region's polygons are rasterized in the
 * rendered view's pixel space and reverse-projected to triangles
 * ([AiMaskProjection], the same path `get_mask_for_text` uses), and
 * those triangles are overwritten with the region's color. Colors are
 * quantized to the perceptually-nearest filament slot via
 * [ColorScience] CIEDE2000 — the same metric M2 / `GamutMatcher` use.
 *
 * Coordinate / index invariants are M2's: camera in
 * `centered_preview_mm`, triangle ids in derived-STL order (==
 * `paintFilamentIndex`). Physical-only (FS deferred, M3).
 */
internal object SemanticPaintPlanner {

    data class RegionStat(val name: String, val slot: Int, val triangleCount: Int)

    data class Resolved(
        /** size = `bvh.triCount`, every entry a 1-based slot (full coverage). */
        val perTriangleSlot: ByteArray,
        val baseSlot: Int,
        val regions: List<RegionStat>,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Resolved) return false
            return perTriangleSlot.contentEquals(other.perTriangleSlot) &&
                baseSlot == other.baseSlot && regions == other.regions
        }

        override fun hashCode(): Int = perTriangleSlot.contentHashCode()
    }

    /**
     * @return null when the plan has no usable `base_color` or the
     *   palette is empty (the caller surfaces an error). Regions whose
     *   color is unparseable or whose polygons are degenerate / project
     *   to nothing are skipped (recorded with triangleCount 0) rather
     *   than failing the whole plan.
     */
    fun resolve(
        bvh: MeshBvh,
        camera: AiRenderEngine.CameraSpec,
        plan: JSONObject,
        palette: List<AutoPaintImageEngine.PaletteSlot>,
    ): Resolved? {
        val n = bvh.triCount
        if (n == 0 || palette.isEmpty()) return null
        if (camera.widthPx <= 0 || camera.heightPx <= 0) return null
        val baseLab = ColorScience.parseLab(plan.optString("base_color").trim()) ?: return null
        val baseSlot = nearestSlot(baseLab, palette)
        val out = ByteArray(n) { baseSlot.toByte() }

        val stats = ArrayList<RegionStat>()
        val regionsArr: JSONArray? = plan.optJSONArray("regions")
        if (regionsArr != null) {
            for (i in 0 until regionsArr.length()) {
                val r = regionsArr.optJSONObject(i) ?: continue
                val name = r.optString("name").trim().ifEmpty { "region_$i" }
                val colLab = ColorScience.parseLab(r.optString("color").trim())
                if (colLab == null) { stats.add(RegionStat(name, 0, 0)); continue }
                val slot = nearestSlot(colLab, palette)
                val polys = parsePolygons(r.optJSONArray("polygons"))
                if (polys.isEmpty()) { stats.add(RegionStat(name, slot, 0)); continue }
                val mask = AiMaskProjection.rasterizePolygons(polys, camera.widthPx, camera.heightPx)
                val tris = AiMaskProjection.project(
                    bvh = bvh, camera = camera, mask = mask,
                    depthMode = AiMaskProjection.DepthMode.FrontFacingOnly,
                    backFaceFilter = true,
                )
                for (t in tris) if (t in 0 until n) out[t] = slot.toByte()
                stats.add(RegionStat(name, slot, tris.size))
            }
        }
        return Resolved(out, baseSlot, stats)
    }

    private fun nearestSlot(
        lab: ColorScience.Lab,
        palette: List<AutoPaintImageEngine.PaletteSlot>,
    ): Int {
        var best = palette[0].slot
        var bestD = Double.POSITIVE_INFINITY
        for (p in palette) {
            val d = ColorScience.ciede2000(lab, p.lab)
            if (d < bestD) { bestD = d; best = p.slot }
        }
        return best
    }

    /** Flat `[x1,y1,x2,y2,...]` float polygons; an entry is rejected if
     *  it isn't an even-length list of ≥3 vertices. Mirrors
     *  AiVisionMaskTools' polygon parsing contract. */
    private fun parsePolygons(arr: JSONArray?): List<FloatArray> {
        if (arr == null) return emptyList()
        val out = ArrayList<FloatArray>(arr.length())
        for (i in 0 until arr.length()) {
            val poly = arr.optJSONArray(i) ?: continue
            val len = poly.length()
            if (len < 6 || len % 2 != 0) continue
            val fa = FloatArray(len) { poly.optDouble(it, Double.NaN).toFloat() }
            if (fa.any { it.isNaN() }) continue
            out.add(fa)
        }
        return out
    }
}

File: web/src/features/SemanticPaintPlanner.ts
import { AiMaskProjection, DepthMode } from './AiMaskProjection';

export interface RegionStat {
  name: string;
  slot: number;
  triangleCount: number;
}

export interface ResolvedPlan {
  perTriangleSlot: Int8Array;
  baseSlot: number;
  regions: RegionStat[];
}

export interface CameraSpec {
  widthPx: number;
  heightPx: number;
  projMatrixRowMajor: Float32Array;
  viewMatrixRowMajor: Float32Array;
}

export interface PaletteSlot {
  slot: number;
  lab: { l: number; a: number; b: number };
}

// Basic stub for ColorScience CIEDE2000
const ciede2000 = (lab1: any, lab2: any) => {
  const dl = lab1.l - lab2.l;
  const da = lab1.a - lab2.a;
  const db = lab1.b - lab2.b;
  return Math.sqrt(dl * dl + da * da + db * db); // simplified fallback
};

const parseLab = (_hex: string) => {
  return { l: 50, a: 0, b: 0 }; // stub
};

export class SemanticPaintPlanner {
  public static resolve(bvh: any, camera: CameraSpec, plan: any, palette: PaletteSlot[]): ResolvedPlan | null {
    const n = bvh.triCount;
    if (n === 0 || palette.length === 0) return null;
    if (camera.widthPx <= 0 || camera.heightPx <= 0) return null;

    const baseColorStr = (plan.base_color || '').trim();
    const baseLab = parseLab(baseColorStr);
    if (!baseLab) return null;

    const baseSlot = this.nearestSlot(baseLab, palette);
    const out = new Int8Array(n);
    out.fill(baseSlot);

    const stats: RegionStat[] = [];
    const regions = plan.regions || [];

    for (let i = 0; i < regions.length; i++) {
      const r = regions[i];
      const name = (r.name || '').trim() || `region_${i}`;
      const colLab = parseLab((r.color || '').trim());
      if (!colLab) {
        stats.push({ name, slot: 0, triangleCount: 0 });
        continue;
      }
      const slot = this.nearestSlot(colLab, palette);
      const polys = this.parsePolygons(r.polygons);
      if (polys.length === 0) {
        stats.push({ name, slot, triangleCount: 0 });
        continue;
      }
      const mask = AiMaskProjection.rasterizePolygons(polys, camera.widthPx, camera.heightPx);
      const tris = AiMaskProjection.project(bvh, camera, mask, DepthMode.FrontFacingOnly, true);
      for (const t of tris) {
        if (t >= 0 && t < n) {
          out[t] = slot;
        }
      }
      stats.push({ name, slot, triangleCount: tris.length });
    }

    return { perTriangleSlot: out, baseSlot, regions: stats };
  }

  private static nearestSlot(lab: any, palette: PaletteSlot[]): number {
    let best = palette[0].slot;
    let bestD = Infinity;
    for (const p of palette) {
      const d = ciede2000(lab, p.lab);
      if (d < bestD) {
        bestD = d;
        best = p.slot;
      }
    }
    return best;
  }

  private static parsePolygons(arr: any[]): Float32Array[] {
    if (!arr || !Array.isArray(arr)) return [];
    const out: Float32Array[] = [];
    for (const poly of arr) {
      if (!Array.isArray(poly)) continue;
      const len = poly.length;
      if (len < 6 || len % 2 !== 0) continue;

      const fa = new Float32Array(len);
      let valid = true;
      for (let i = 0; i < len; i++) {
        const v = parseFloat(poly[i]);
        if (isNaN(v)) {
          valid = false;
          break;
        }
        fa[i] = v;
      }
      if (valid) out.push(fa);
    }
    return out;
  }
}
